src/api/runner.ts

```diff
@@ -0,0 +1,272 @@
+import * as vscode from "vscode";
+import _ from "lodash";
+import { basename, normalize, relative } from "pathe";
+import stripAnsi from "strip-ansi";
+import ErrorStackParser from "error-stack-parser";
+import type { TestError } from "testplane";
+
+import { TestTree } from "../test-tree";
+import { getUniqBrowserId } from "../test-tree/data";
+import { TestCaseByBrowser, getTestData } from "../test-tree/data";
+import { logger } from "../utils";
+import type {
+    TestplaneMasterRPC,
+    TestplaneMasterHandlers,
+    TestplaneMasterTest,
+    TestplaneMasterTestResult,
+    TestplaneMasterRunTestsOpts,
+} from "../types";
+
+interface LocationFromStack {
+    path: string;
+    line: number;
+    column: number;
+}
+
+enum TestStatuses {
+    BEGIN = "begin",
+    PASS = "pass",
+    FAIL = "fail",
+    PENDING = "pending",
+    ENQUEUED = "enqueued",
+}
+
+export class TestRunner extends vscode.Disposable {
+    private _testRun: vscode.TestRun | undefined;
+
+    constructor(
+        private readonly _controller: vscode.TestController,
+        private readonly _tree: TestTree,
+        private readonly _workspaceFolder: vscode.WorkspaceFolder,
+        private readonly _api: TestplaneMasterRPC,
+        private readonly _handlers: TestplaneMasterHandlers,
+    ) {
+        super(() => {
+            logger.debug?.("Dispose test runner");
+
+            this._handlers.clearListeners();
+            this._endTestRun();
```

Comment:
are we trying to end runner to be sure that there are definitely no other launches?